src/gui/src/layoutViewer.cpp

```diff
@@ -702,15 +732,122 @@ void LayoutViewer::drawRows(dbBlock* block,
 
     dbRowDir dir = row->getDirection();
     int count = row->getSiteCount();
-    for (int i = 0; i < count; ++i) {
-      painter->drawRect(QRect(QPoint(x, y), QPoint(x + w, y + h)));
+    if (w < min_resolution) {
+      // individual sites not visible, just draw the row
       if (dir == dbRowDir::HORIZONTAL) {
-        x += spacing;
-      } else {
-        y += spacing;
+        w = spacing*count;
+      }
+      else {
+        h = spacing*count;
+      }
+      count = 1;
+    }
+    if (h >= min_resolution) {
+      // row height can be seen
+      for (int i = 0; i < count; ++i) {
+        const Rect row(x, y, x + w, y + h);
+        if (row.intersects(bounds)) {
+          // only paint rows that can be seen
+          painter->drawRect(QRect(QPoint(x, y), QPoint(x + w, y + h)));
+        }
+
+        if (dir == dbRowDir::HORIZONTAL) {
+          x += spacing;
+        } else {
+          y += spacing;
+        }
+      }
+    }
+    else {
+      const Rect row(x, y, x + w, y + h);
+      if (row.intersects(bounds)) {
+        rows_for_outline.push_back(row);
       }
     }
   }
+  if (!rows_for_outline.empty()) {
+    // draw outline of rows
+    std::map<Point, int> row_points_counter;
+    for (const Rect& row : rows_for_outline) {
+      row_points_counter[row.ll()]++;
+      row_points_counter[row.lr()]++;
+      row_points_counter[row.ur()]++;
+      row_points_counter[row.ul()]++;
+    }
+
+    // find corners of row blocks
+    std::set<Point> corners;
+    for (const auto& [pt, count] : row_points_counter) {
+      if (count == 1) {
+        corners.insert(pt);
+      }
+    }
+
+    QPolygon outline;
+    do {
+      outline = getRowOutline(corners);
+      painter->drawPolygon(outline);
+    } while (!outline.isEmpty());
+  }
```

Comment:
I don't think a QPolygon can have holes so I'm not clear how this would work.

Reply:
you are correct as far as I can tell. The `getRowOutline` is building a QPolygon for each set of corners that form a complete path. So each QPolygon would just be a complete path (outline) and then several would be drawn if there are holes in the rows. That way there is no need to deal with holes here.
As `getRowOutline` uses the corner points they are removed from the `corners` set.

Does that help?

Reply:
I should also mention it's not a "must" have feature that the row outlines are drawn, so if you think it's hokey, I can just revert those changes out (should only be two commits now). I think it looks nice, but certainly not wed to this either.

Reply:
I like the idea but I'm concerned that the implementation is rather tricky - I am not sure it always works but I can't disprove it either.  It also seems expensive to put in the draw loop - I think I would cache the result since rows rarely change.

Reply:
I agree, I reverted the outline code, so now it just doesn't draw any of the rows until 5x the resolution limit.